src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -406,6 +414,135 @@ class CDCL {
+
+    /**
+     * Preprocessing is ran before the main loop of the solver,
+     * allowing to spend some time simplifying the problem
+     * in exchange for a possibly faster solving time.
+     *
+     * @return if the solution is conclusive after preprocessing,
+     * return the model, otherwise return null
+     */
+    fun preprocess(): Model? {
+        require(level == 0)
+
+        // Don't bother with anything if there is already
+        // a level 0 conflict
+        propagate()?.let { return Model.UNSAT }
+
+        failedLiteralProbing()?.let { return it }
+
+        // Without this, we might let the solver propagate nothing
+        // and make a decision after all values are set.
+        if (allAssignedAndSat()) {
+            return getModel()
+        }
+
+        return null
+    }
+
+    /**
+     * Literal probing is only useful if the probe can lead
+     * lead to derivation of something by itself. We can
+     * generate list of possibly useful probes ahead of time.
+     *
+     * TODO: Right now, this implementation is very naive. To
+     *       If there is a cycle in binary implication graph,
+     *       we will generate a lot of useless probes that will
+     *       propagate in exactly the same way and possibly
+     *       produce a lot of duplicate/implied binary clauses.
+     *       This will be fixed with the implementation of
+     *       Equivalent Literal Substitution (ELS).
+     *
+     *  @return list of literals to try probing with
+     */
+    private fun generateProbes(): List<Lit> {
+        val probes = mutableSetOf<Lit>()
+
+        for (clause in constraints) {
+            // (A | B) <==> (-A -> B) <==> (-B -> A)
+            // Both -A and -B can be used as probes
+            if (clause.size == 2) {
+                val (a, b) = clause
+                if (getValue(a) == LBool.UNDEFINED) probes.add(a.neg)
+                if (getValue(b) == LBool.UNDEFINED) probes.add(b.neg)
+            }
+        }
+
+        println(probes.size)
+        return probes.toMutableList()
+    }
+
+    /**
+     * Try to propagate each probe and see if it leads to
+     * deduction of new binary clauses, or maybe even a
+     * conflict.
+     *
+     * Consider clauses (-1, 2), (-1, -2, 3), and a probe 1.
+     * By propagating 1, we can derive 2, which in turn
+     * allows us to derive 3. This means that we can add
+     * a new binary clause (-1, 3) to the problem. There are
+     * other mechanisms that can derive this clause, but
+     * probing is one of them.
+     */
+    private fun failedLiteralProbing(): Model? {
+        val probesToTry = generateProbes()
+
+        for (probe in probesToTry) {
+            // If we know that the probe is already true, just skip it.
+            if (getValue(probe) == LBool.TRUE) {
+                continue
+            }
+
+            level++
+            uncheckedEnqueue(probe)
+            val startOfProbeTrail = trail.size
```

Comment:
The variable `startOfProbeTrail` does not have an obvious meaning from its name. Maybe refer to the size of the trail in the variable's name?